node/src/sub_lib/neighborhood.rs

```diff
@@ -115,105 +119,123 @@ impl NeighborhoodMode {
-    fn from(tuple: (&NodeRecord, bool, &dyn CryptDE)) -> Self {
-        let (node_record, mainnet, cryptde) = tuple;
+impl From<(&NodeRecord, Blockchain, &dyn CryptDE)> for NodeDescriptor {
+    fn from(tuple: (&NodeRecord, Blockchain, &dyn CryptDE)) -> Self {
+        let (node_record, blockchain, cryptde) = tuple;
         NodeDescriptor {
+            blockchain,
             encryption_public_key: cryptde
                 .descriptor_fragment_to_first_contact_public_key(
                     &cryptde.public_key_to_descriptor_fragment(node_record.public_key()),
                 )
                 .expect("Internal error"),
-            mainnet,
             node_addr_opt: node_record.node_addr_opt(),
         }
     }
 }
 
 impl NodeDescriptor {
-    pub fn from_str(cryptde: &dyn CryptDE, s: &str) -> Result<NodeDescriptor, String> {
-        let (mainnet, pieces) = {
-            let chain_id = chain_id_from_name(DEFAULT_CHAIN_NAME);
-            let delimiter = node_descriptor_delimiter(chain_id);
-            let pieces: Vec<&str> = s.splitn(2, delimiter).collect();
-            if pieces.len() == 2 {
-                (true, pieces)
-            } else {
-                let chain_id = chain_id_from_name("ropsten");
-                let delimiter = node_descriptor_delimiter(chain_id);
-                let pieces: Vec<&str> = s.splitn(2, delimiter).collect();
-                if pieces.len() == 2 {
-                    (false, pieces)
-                } else {
-                    return Err(format!(
-                        "Should be <public key>[@ | :]<node address>, not '{}'",
-                        s
-                    ));
-                }
-            }
-        };
-
-        let encryption_public_key =
-            match cryptde.descriptor_fragment_to_first_contact_public_key(pieces[0]) {
-                Err(e) => return Err(e),
-                Ok(hpk) => hpk,
-            };
-
-        let node_addr_opt = {
-            if pieces[1] == ":" {
-                None
-            } else {
-                match NodeAddr::from_str(pieces[1]) {
-                    Err(e) => return Err(e),
-                    Ok(node_addr) => Some(node_addr),
-                }
-            }
+    pub fn from_str(cryptde: &dyn CryptDE, str_descriptor: &str) -> Result<NodeDescriptor, String> {
+        let (blockchain, key, str_node_addr) = Self::try_dismantle_str(str_descriptor)?;
+        let encryption_public_key = cryptde.descriptor_fragment_to_first_contact_public_key(key)?;
+        let node_addr_opt = if str_node_addr == ":" {
+            None
+        } else {
+            Some(NodeAddr::from_str(str_node_addr)?)
         };
 
         Ok(NodeDescriptor {
+            blockchain,
             encryption_public_key,
-            mainnet,
             node_addr_opt,
         })
     }
 
+    fn try_dismantle_str(str_descriptor: &str) -> Result<(Blockchain, &str, &str), String> {
```

Comment:
Would a more common verb used in a scenario like this be `parse`? `try_parse() -> Option(Blockchain)` or `parse() -> Option(Uri)`